motor/core.pyi

```diff
@@ -0,0 +1,782 @@
+        max_await_time_ms: Optional[int] = None,
+        batch_size: Optional[int] = None,
+        collation: Optional[_CollationIn] = None,
+        start_at_operation_time: Optional[Timestamp] = None,
+        session: Optional[AgnosticClientSession] = None,
+        start_after: Optional[Mapping[str, Any]] = None,
+        comment: Optional[str] = None,
+        full_document_before_change: Optional[str] = None,
+        show_expanded_events: Optional[bool] = None,
+    ) -> AgnosticChangeStream: ...
+    def __getattr__(self, name: str) -> Any: ...
+    def __getitem__(self, name: str) -> Any: ...
+    def wrap(self, obj: Any) -> Any: ...
+
+class _MotorTransactionContext:
+    _session: AgnosticClientSession
+
+    def __init__(self, session: AgnosticClientSession): ...
+    async def __aenter__(self) -> _MotorTransactionContext: ...
+    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None: ...
+
+class AgnosticClientSession(AgnosticBase):
+    __motor_class_name__: str
+    __delegate_class__: Type[ClientSession]
+
+    async def commit_transaction(self) -> None: ...
+    async def abort_transaction(self) -> None: ...
+    async def end_session(self) -> None: ...
+    def cluster_time(self) -> Optional[Mapping[str, Any]]: ...
+    def has_ended(self) -> bool: ...
+    def in_transaction(self) -> bool: ...
+    def options(self) -> SessionOptions: ...
+    def operation_time(self) -> Optional[Timestamp]: ...
+    def session_id(self) -> Mapping[str, Any]: ...
+    def advance_cluster_time(self, cluster_time: Mapping[str, Any]) -> None: ...
+    def advance_operation_time(self, operation_time: Timestamp) -> None: ...
+    def __init__(self, delegate: ClientSession, motor_client: AgnosticClient): ...
+    def get_io_loop(self) -> Any: ...
+    async def with_transaction(
+        self,
+        coro: Callable[..., Coroutine[Any, Any, Any]],
+        read_concern: Optional[ReadConcern] = None,
+        write_concern: Optional[WriteConcern] = None,
+        read_preference: Optional[_ServerMode] = None,
+        max_commit_time_ms: Optional[int] = None,
+    ) -> _T: ...
+    def start_transaction(
+        self,
+        read_concern: Optional[ReadConcern] = None,
+        write_concern: Optional[WriteConcern] = None,
+        read_preference: Optional[_ServerMode] = None,
+        max_commit_time_ms: Optional[int] = None,
+    ) -> _MotorTransactionContext: ...
+    @property
+    def client(self) -> AgnosticClient: ...
+    async def __aenter__(self) -> AgnosticClientSession: ...
+    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None: ...
+    def __enter__(self) -> None: ...
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None: ...
+
+class AgnosticDatabase(AgnosticBaseProperties):
+    __motor_class_name__: str
+    __delegate_class__: Type[Database]
+
+    def __hash__(self) -> int: ...
+    def __bool__(self) -> int: ...
+    async def command(
```

Comment:
We should use the overload pattern for command() now: https://github.com/mongodb/mongo-python-driver/blob/e0b8b36f4162eccd0b0fbe00563a749e90ee37d6/pymongo/database.py#L758-L800

Reply:
I'm running into issues when trying to use the `@overload` pattern, it seems that the type system is resolving `CodecOptions[_CodecDocumentType]` as `Any` in the second signature:

```
motor/core.pyi:264: error: Overloaded function signature 3 will never be matched: signature 2's parameter type(s) are the same or broader  [misc]
        async def command(
        ^
```

Reply:
Hmm could you then back this change out and open a new ticket for it explaining this problem?

Reply:
Created https://jira.mongodb.org/browse/MOTOR-1173.